`spack info`: show conditional dependencies and licenses; allow filtering

tgamblin opened the pull request:
`spack info` has always shown dependencies grouped by type, but not conditional dependency information. While conditional deps are often very relevant, users have to read raw package files to see them. 

This PR generifies the code used to display variants and uses it for dependencies and licenses too. As with variants, dependencies are shown grouped by conditions, with unconditional dependencies first.  Users can supply the `--by-name` argument to list ordered first by name instead.

`--by-name` is meant to replace `--variants-by-name`. `--variants-by-name` still works but is suppressed from help and has the same function as `--by-name`.

`spack info gcc` before and after:

<img width="200" height="195" valign="top" alt="Screenshot 2025-08-10 at 10 28 36 PM" src="https://github.com/user-attachments/assets/5a20e90c-f485-4fd1-b833-9e0e49e5ebc2" />

<img width="200" height="1013" alt="Screenshot 2025-08-10 at 10 29 20 PM" src="https://github.com/user-attachments/assets/e72af675-e1ed-43f9-aa06-0e5138c417c2" />

As there is increasingly more information displayed by `spack info`, it can be useful to filter out variants, dependencies, versions, etc. that are irrelevant to certain package configurations.

This PR also allows `spack info` to take a spec instead of a package name, so users can run, e.g.:

```console
spack info hdf5@1.14
```

to get *only* information relevant to 1.14.x versions of `hdf5`.  Versions, dependencies, variants, and licenses are now filtered based on the input spec.

Before and after:

<img width="200" valign="top" height="429" alt="Screenshot 2025-08-10 at 10 33 03 PM" src="https://github.com/user-attachments/assets/b4401489-ab0b-4bea-8db3-bc01b37fc4b7" />

<img width="200" valign="top" height="302" alt="Screenshot 2025-08-10 at 10 33 22 PM" src="https://github.com/user-attachments/assets/087b6087-b4bd-42a0-a3d3-327decd097e6" />

- [x] generify when-grouping code to use a `Formatter` class
- [x] Rewrite variant code to use generic formatter.
- [x] Rewrite dependency code to use generic formatter.
- [x] Rewrite license code to use generic formatter.
- [x] add full type annotations to `cmd/info.py`
- [x] allow `spack info` to take a spec instead of a package name
- [x] filter `spack info` output versions, dependencies, variants, and licenses based on the input spec
- [x] tests

tgamblin committed:
`spack info`: show full dependency information

`spack info` has always shown dependencies grouped by type, but not
conditional dependency information.  This generifies the code used
to display variants and uses it for dependencies as well.

As with variants, dependencies are shown grouped by conditions, with
unconditional dependencies first.

- [x] generify when-grouping code to use a `Formatter` class
- [x] Rewrite variant code to use generic formatter.
- [x] Rewrite dependency code to use generic formatter.
- [x] add full type annotations to `cmd/info.py`

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: use when formatting for licenses

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: allow filtering by when

`spack info` has historically accepted a package name but not a spec.

Now that it shows lots of conditional information, it can be useful
to filter out variants, dependencies, versions, etc. that are irrelevant
to certain package configurations.

- [x] allow `spack info` to take a spec instead of a package name
- [x] filter `spack info` output versions, dependencies, variants, and
      licenses based on the input spec

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo reviewed (commented):
Overall this looks great. I think there is an issue with single-value variant. For instance check the dependencies of:
```console
$ spack info zlib-ng build_system=autotools
```
vs.
```console
$ spack info zlib-ng build_system:=autotools
```
Since `build_system` is `multi=False` specifying a value prevents other values to be possible.

Another issue is with conditional variants. For instance, in all `CudaPackage`s, if we specify:
```console
$ spack info trilinos~cuda
```
we continue to see pages of `cuda_arch`, even if that variant is gated on `+cuda`. 

Finally, just a comment on colors (disregard this if you don't agree). I wouldn't use red for run dependencies, cause red is usually associated with failures. Maybe yellow?
<div align="center">
<img width="416" height="329" alt="Screenshot from 2025-08-15 09-36-52" src="https://github.com/user-attachments/assets/ad271a28-7412-4d43-8318-3dfb2e037516" />
</div>
I'll dive into code next.

alalazo commented on lib/spack/spack/package_base.py:
```diff
@@ -447,7 +447,7 @@ def _precedence(obj) -> int:
     if precedence is None:
-        raise KeyError(f"Couldn't get precedence from {type(obj)}")
+        return 0  # raise KeyError(f"Couldn't get precedence from {type(obj)}")
     return precedence
```
It seems this can become:
```python
return getattr(obj, "precedence", 0)
```
Probably it can also be inlined.

tgamblin replied:
done.

alalazo commented on lib/spack/spack/test/cmd/info.py:
```diff
@@ -85,8 +50,29 @@ def test_info_fields(pkg_query, extra_args, parser, print_buffer):
+        (False, ["bowtie@1.3:"], ["1.4.0", "1.3.0"], ["1.2.2", "1.2.0"]),
+        (False, ["bowtie@1.2"], ["1.2.2", "1.2.0"], ["1.3.0"]),  # 1.4.0 still shown as preferred
+    ],
+)
```
If, as it seems, we expect no differences between tests using `--by-name` vs. not, we can shorten the definition like:
```suggestion
    "args,in_output,not_in_output",
    [
        (["licenses-1 +foo"], ["MIT"], ["Apache-2.0"]),
        (["licenses-1 ~foo"], ["Apache-2.0"], ["MIT"]),
        (["bowtie"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
        (["bowtie@1.2:"], ["1.4.0", "1.3.0", "1.2.2", "1.2.0"], []),
        (["bowtie@1.3:"], ["1.4.0", "1.3.0"], ["1.2.2", "1.2.0"]),
        (["bowtie@1.2"], ["1.2.2", "1.2.0"], ["1.3.0"]),  # 1.4.0 still shown as preferred
    ],
)
@pytest.mark.parametrize("by_name", [True, False])
```

alalazo replied:
It would be good to also test single valued variants, and conditional variants. If there are cases where we would like to have a result, but it's difficult to infer it, I think it's fine to document the case as xfail.

tgamblin replied:
done in a testing commit I'm going to push

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -463,26 +548,12 @@ def print_virtuals(pkg, args):
+    if len(specs) > 1:
+        tty.die(f"spack info requires exactly one spec. Parsed {len(specs)}")
 
-def info(parser, args):
-    spec = spack.spec.Spec(args.package)
+    spec = specs[0]
     pkg_cls = spack.repo.PATH.get_pkg_class(spec.fullname)
     pkg = pkg_cls(spec)
```
Here we probably lack some substitution that would make the single-value variant work. With a debugger I see:
```python
>>> pkg.spec.variants["build_system"].type
<VariantType.MULTI: 3>
```
which indicates it's treated as a multi-value variant.

tgamblin replied:
This should be addressed.

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -263,146 +253,241 @@ def print_tests(pkg, args):
+
+def print_variants(pkg: PackageBase, args: Namespace) -> None:
+    """output variants"""
+    print_definitions(pkg, "Variants", pkg.variants, VariantFormatter(), args.by_name)
```
For the `trilinos~cuda` issue (which is affecting all ROCm and CUDA packages):
1. We correctly infer that `cuda_arch` _is not_ a possible variant in the "Variants" section, but
2. We don't store this information and show a lot of conditional dependencies on `cuda` gated on `when="cuda_arch=..."`

If we store the set of possible variants, we can also filter dependencies.

tgamblin replied:
This should be addressed.

alalazo reviewed (changes requested):
Minor comments, and two hints on the variant issues.

haampie commented:
Does this fix #51222?

tgamblin commented:
I would say that it is in the spirit of #51222, in that it allows filtering by spec, but it doesn't exactly fix it b/c it doesn't address filtering by things like `--phases` or other metadata that's not on the spec.

I think a refactor towards something more general like `--show versions,phases` is in order (as there is a constantly growing number of sections). The current args are inconsistent as some of them are positive (`--phases`) and some of them are negative (`--no-dependencies`). I would rather just let the user say what they want.

tgamblin commented:
Ok @alalazo you gotta help me on this one - I apparently wasn't in the loop with you and @haampie on the review of #35681, but in testing this I noticed that [this code](https://github.com/spack/spack/pull/35681/files#diff-a8f340acf63e6506e774ad8ee1b0c491bf325ef6dada3b06ad28d6e158c4eeaaR1526) breaks the `spack info` filtering here for variant and dependency conditions that include versions, because `spack.fetch_strategy.for_package_version()` is actually modifying `pkg.spec`.

Does it have to do that?  Can we not defensively copy in `spack.fetch_strategy.for_package_version()`?  Why does it need to mutate the spec on the passed in package?

tgamblin committed:
fetch_strategy: for_package_version should not mutate `pkg.spec`

This is a hack while we figure out why it's doing this

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
Add `PackageBase.intersects()` to include more static information

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: make `--by-name` the default, `--by-when` an option

After adding dependencies, it is often much clearer to list info output by name, since
there can be a very large number of dependency conditions in the output.

`--by-when` is still an option, and the `--by-name` argument is still present even
though it is the default.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: fix error messages when 1 or 0 specs are provided

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: validate variant names in input specs

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: suggest disabling variants when there are too many conditions

This will advise users to add, e.g., `~cuda` and `~rocm` when output for a package is
very long. This makes the feature more discoverable.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: make text wrapping in output color-aware

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: ensure varant values / deptypes have minimum width

For complex dependencies like boost, there may be many required variants. Ensure that
they don't push deptypes off the edge of the screen by setting a min width and showing
deptypes on the next line when the dependency spec is too long.

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin committed:
`spack info`: show compiler info in dependency specs

- need to use long_spec for this

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

tgamblin commented:
@alalazo: This PR is getting fairly large now but it's much improved. The commits are, I think, well factored into independent fixes/features now.

1. Fixing all the spec stuff above requires knowing something about the package. I was using `Spec.intersects()`, which says whether two constraints *may* coincide, but it knows too little about variants to do what you want, e.g., to know that a variant is only ever single-valued, or to know that the presence of a variant that's conditional on the value of some single-valued variant implies that no other values for that variant are allowable. I put in some static analysis in a new method on the `Package` to handle these cases, and it allows for much more aggressive filtering.  e.g., if you ask for `generator=make`, it knows to rule out dependencies with `build_system=autotools`.
2. I added something that's maybe controversial -- if there are lots of deps, Spack now recommends getting rid of the bigger clumps of them by toggling certain variants at the end of `spack info`. I may remove this reminder later, but for now it's there to help with the many packages that use loops to set up locked `cuda` and/or `rocm` dependencies.  I think this will ultimately be made much less necessary by #41700, (where lots of rules will be reduced to something like `depends_on("dep cuda_arch={self.variants['cuda_arch'].value}"), but that is not ready yet.
3. I made text wrapping in `spack info` color-aware, which saves a space and avoids weird formatting.
4. I added some code to handle cases where the dependency spec is very long, e.g. some dependencies on `boost` have many variants. Instead of cramming the deptypes against the right margin, one letter per line, `spack info` now reserves 30 cols for the values in variant definitions, and it just puts them on the next line if the dep spec is too long.

See what you think and let me know if this is more usable.

tgamblin committed:
`spack info`: add more extensive testing for filtering

Signed-off-by: Todd Gamblin <tgamblin@llnl.gov>

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -56,108 +123,97 @@ def setup_parser(subparser: argparse.ArgumentParser) -> None:
 
-    arguments.add_common_arguments(subparser, ["package"])
+    # deprecated for the more generic --by-name, but still here until we can remove it
+    subparser.add_argument(
+        "--variants-by-name", dest="by_name", action="store_true", help=argparse.SUPPRESS
+    )
```
(Question) Should we also print a deprecation warning?

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -56,108 +123,97 @@ def setup_parser(subparser: argparse.ArgumentParser) -> None:
+    big_variants = [
+        (name, val)
+        for n, (name, val) in variant_counts
+        # counts >= 40 and skip if the user already disabled the variant
```
(nitpick) Let's use something generic in comments, so they don't go out of sync
```suggestion
        # counts >= <threshold> and skip if the user already disabled the variant
```

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -263,146 +319,258 @@ def print_tests(pkg, args):
+        name_field: name and optional info, e.g. a default; should be short.
+        values_field: possible values for the entry; Wrapped if long.
+        description: description of the field (wrapped if overly long)
+        max_name_len:
```
(niptpick) Missing description, although it's kind of clear by the name 🙂

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -263,146 +319,258 @@ def print_tests(pkg, args):
 
-def _print_variants_header(pkg):
-    """output variants"""
+K = TypeVar("K", bound=SupportsRichComparison)
+V = TypeVar("V")
```
Do we use these two variables?

alalazo commented on lib/spack/spack/test/cmd/info.py:
```diff
@@ -2,77 +2,51 @@
 @pytest.mark.parametrize("extra_args", [[], ["--variants-by-name"]])
 def test_it_just_runs(extra_args):
-    info("vtk-m", *extra_args)
+    info(*extra_args, "vtk-m")
```
(minor) Not a fault of this PR, so can be addressed later, but I wonder if these kind of tests are meaningful, or should be deleted. Here there is no use case, or assertion.

alalazo commented on lib/spack/spack/package_base.py:
```diff
@@ -1458,6 +1463,65 @@ def provides(self, vpkg_name: str) -> bool:
             if self.spec.intersects(when_spec)
         )
 
+    def intersects(self, spec: spack.spec.Spec) -> bool:
```
👍

alalazo reviewed (approved):
The use cases pointed out in the previous review have been addressed 💯 Only minor comments that we can leave for a following PR.
